docs/proposals/kubelet-eviction.md

```diff
@@ -0,0 +1,200 @@
+--eviction-soft-period="": Comma-delimited list of eviction periods for each soft eviction signal.  For example, 'memory.available=30s'
+```
+
+If no eviction period is configured for a corresponding soft eviction threshold,
+it will default to 30s for each soft eviction signal.
+
+An eviction expression is of the form:
+
+`<eviction-signal><operator><quantity>`
+
+* valid `eviction-signal` tokens are `memory.available` (future signals are anticipated `storage` or `cpu.load`)
+* valid `operator` tokens are `<` and `>`
+* valid `quantity` tokens must match the quantity representation used by kubernetes
+
+
+## Eviction Monitoring Interval
+
+The `kubelet` will support monitoring available compute resource usage at the same
+specified housekeeping interval as `cAdvisor` housekeeping (10s in v1.2 by default).
+
+## Eviction scenario
+
+Let's assume the operator started the `kubelet` with the following:
+
+```
+--eviction-hard="memory.available<100Mi"
+--eviction-soft="memory.available<300Mi"
+--eviction-soft-period="memory.available=30s"
+```
+
+The `kubelet` will run a sync loop that looks at the available memory
+on the node as reported from `cAdvisor` by calculating (capacity - workingSet).
+If available memory is observed to drop below 100Mi, the `kubelet` will immediately
+initiate eviction. If available memory is observed as falling below `300Mi`,
+it will record when that signal was observed internally in a cache.  If at the next
+sync, that criteria was no longer satisfied, the cache is cleared for that
+signal.  If that signal is observed as being satisfied for longer than the
+specified period, the `kubelet` will initiate eviction to attempt to
+reclaim the resource that has met its eviction threshold.
+
+## Eviction of Pods
+
+If an eviction threshold has been met, the `kubelet` will initiate the
+process of evicting pods until it has observed the signal has gone below
+its defined threshold.
+
+The eviction sequence works as follows:
+
+* for each monitoring interval, if eviction thresholds have been exceeded
+ * find candidate pod
+ * fail the pod
+ * wait until pod is terminated
+
+This means that there is no parallel killing of pods, and we will wait
+a single monitoring interval before killing an additional pod to allow
+for reclamation of the starved resource to take place.  By default, the
+`kubelet` will respect graceful termination period when doing eviction.
+
+The mechanism of finding a candidate pod for eviction is tied to an
+eviction strategy.  An eviction strategy is a compiled in plug-in that
+delegates finding a candidate pod.  The `kubelet` is configured to work
+against a particular strategy.  The default strategy if not specified is
+`greedy`.  Other strategies are anticipated in the future.
+
+```
+--eviction-strategy="greedy": Strategy to use when eviction thresholds have been met.
+```
+
+### Eviction Strategy: greedy
+
+The `greedy` strategy targets pods that are the largest consumers of the
```

Comment:
Given that we are adding `Priority` to pods, should we even support multiple policies? [Pod Priority proposal](https://github.com/kubernetes/kubernetes/pull/22217/files#diff-8141aabac2eced92d5560d2d24c1a2d7R143)

The current `greedy` policy would then evolve into a `default` policy.